File: Leetcode100/leetcode100_python/BinaryTreeMaximumPathSum.py
class TreeNode:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        self.result = float('-inf')
        def helper(root):
            if not root:
                return 0
            left_max = max(helper(root.left), 0)
            right_max = max(helper(root.right), 0)
            self.result = max(self.result, left_max + root.val + right_max)
            return max(left_max, right_max) + root.val
        if not root:
            return 0
        helper(root)
        return self.result

if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    root = TreeNode(nums[0])
    queue = [root]
    for i in range(1, len(nums), 2):
        cur = queue.pop(0)
        if nums[i] != -1:
            cur.left = TreeNode(nums[i])
            queue.append(cur.left)
        if nums[i + 1] != -1:
            cur.right = TreeNode(nums[i + 1])
            queue.append(cur.right)
    solution = Solution()
    print(solution.maxPathSum(root))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/binary_tree_maximum_path_sum/BinaryTreeMaximumPathSum.java
package main.java.com.jiang.leetcode100.binary_tree_maximum_path_sum;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.Scanner;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) {this.val = val;}
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public int result = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        if(root == null) return 0;
        helper(root);
        return result;
    }

    public int helper(TreeNode root) {
        if(root == null) return 0;
        int left_sum = Math.max(helper(root.left), 0);
        int right_sum = Math.max(helper(root.right), 0);
        result = Math.max(result, left_sum + root.val + right_sum);
        return Math.max(left_sum, right_sum) + root.val;
    }
}

public class BinaryTreeMaximumPathSum {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] nums = new  int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            nums[i] = Integer.parseInt(nums_str[i]);
        }

        TreeNode root = new TreeNode(nums[0], null, null);
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);

        for(int i = 1; i < nums.length; i += 2) {
            TreeNode cur = queue.poll();
            if(nums[i] != -1) {
                cur.left = new TreeNode(nums[i], null, null);
                queue.add(cur.left);
            }
            if(nums[i + 1] != -1) {
                cur.right = new TreeNode(nums[i + 1], null, null);
                queue.add(cur.right);
            }
        }

        System.out.println(new Solution().maxPathSum(root));
    }
}